File: src/test/java/top/cnuo/idlepool/util/LanguageFilesTest.java
package top.cnuo.idlepool.util;

import org.junit.jupiter.api.Test;
import org.yaml.snakeyaml.Yaml;

import java.io.InputStream;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

class LanguageFilesTest {
    @Test
    void englishAndChineseLanguageFilesHaveMatchingKeys() {
        Map<String, Object> chinese = load("message.yml");
        Map<String, Object> english = load("en.yml");

        Set<String> chineseKeys = new LinkedHashSet<>();
        Set<String> englishKeys = new LinkedHashSet<>();
        flatten("", chinese, chineseKeys);
        flatten("", english, englishKeys);

        assertEquals(chineseKeys, englishKeys, "message.yml and en.yml must expose the same message keys");
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> load(String name) {
        InputStream stream = LanguageFilesTest.class.getClassLoader().getResourceAsStream(name);
        assertNotNull(stream, "Missing language resource: " + name);
        return new Yaml().load(stream);
    }

    @SuppressWarnings("unchecked")
    private static void flatten(String prefix, Map<String, Object> section, Set<String> keys) {
        section.forEach((name, value) -> {
            String path = prefix.isEmpty() ? name : prefix + "." + name;
            if (value instanceof Map<?, ?> nested) {
                flatten(path, (Map<String, Object>) nested, keys);
            } else {
                keys.add(path);
            }
        });
    }
}

File: src/test/kotlin/top/cnuo/idlepool/util/LanguageFilesTest.kt
package top.cnuo.idlepool.util

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import org.yaml.snakeyaml.Yaml

class LanguageFilesTest {
    @Test fun `english and chinese expose matching keys`() {
        val chinese = mutableSetOf<String>(); val english = mutableSetOf<String>()
        flatten("", load("message.yml"), chinese); flatten("", load("en.yml"), english)
        assertEquals(chinese, english, "message.yml and en.yml must expose the same message keys")
    }
    private fun load(name: String): Map<String, Any?> {
        val stream = javaClass.classLoader.getResourceAsStream(name); assertNotNull(stream, "Missing language resource: $name")
        @Suppress("UNCHECKED_CAST") return Yaml().load(stream) as Map<String, Any?>
    }
    private fun flatten(prefix: String, section: Map<String, Any?>, keys: MutableSet<String>) {
        section.forEach { (name, value) ->
            val path = if (prefix.isEmpty()) name else "$prefix.$name"
            if (value is Map<*, *>) { @Suppress("UNCHECKED_CAST") flatten(path, value as Map<String, Any?>, keys) } else keys += path
        }
    }
}
